device-server/src/main/kotlin/com/badoo/automation/deviceserver/host/AndroidEmulatorsNode.kt

```diff
@@ -0,0 +1,352 @@
+                    return
+                }
+            }
+    }
+
+    private fun summary(body: String?, response: Response): String {
+        return  if (body != null && body.isNotEmpty()) body else response.toString()
+    }
+
+    val nullRequestBody = RequestBody.create(null, "")
+
+    private fun httpPost(endpoint: String): Response {
+        return httpClient.newCall(buildEndpoint(endpoint).post(nullRequestBody).build()).execute()
+    }
+
+    private fun httpGet(endpoint: String): Response {
+        return httpClient.newCall(buildEndpoint(endpoint).build()).execute()
+    }
+
+    private fun buildEndpoint(endpoint: String) = Request.Builder().url("http://$ticketHost/$endpoint")
+
+    private fun throwUnlessOK(response: Response) {
+        if (response.code() != HttpURLConnection.HTTP_OK) {
+            throw DeviceCreationException("NOT HTTP_OK: $response")
+        }
+    }
+
+    private fun releaseOnTicketMachine(ref: DeviceRef): Boolean {
+        httpPost("docker/release/$ref")
+                .use {
+                    throwUnlessOK(it)
+                    val body = it.body()?.string()
+                    val released = body?.contains(":SUCCESS: $ref released") ?: false
+                    if (!released) {
+                        logger.warn("Unable to release $ref: $body $it")
+                    }
+                    return released
+                }
+    }
+
+    private fun emulatorOrNullFromRef(ref: DeviceRef): EmulatorSerial? {
+        val deviceDTO = refToDto[ref]
+        if (deviceDTO == null) {
+            logger.warn("No emulator for $ref")
+            return null
+        }
+        return deviceDTO.info.udid
+    }
+
+    private fun updateDtoForRef(ref: String, state: DeviceState, emulatorSerial: EmulatorSerial,
+                                adb_socket: URI = URI(""), model: String,
+                                last_error: Throwable? = null): DeviceDTO {
+        val deviceDTO = DeviceDTO(
+                ref = ref,
+                state = state,
+                fbsimctl_endpoint = URI(ticketHost),
+                wda_endpoint = adb_socket,
+                info = DeviceInfo(udid = emulatorSerial, model = model, os = "", arch = "", name = ""),
+                calabash_port = 0, user_ports = HashSet(), capabilities = null, last_error = last_error?.toDto())
+        refToDto[ref] = deviceDTO
+        return deviceDTO
+    }
+
+    private fun deallocateEmulatorSlot(emulator: EmulatorSerial): DeviceRef? {
+        return allocatedEmulators.remove(emulator)
+    }
+
+    private fun allocateEmulatorSlot(ref: DeviceRef): EmulatorSerial {
+        synchronized(allocatedEmulators) {
+            try {
+                val emulator =
+                        (knownEmulators - allocatedEmulators.keys).asSequence().sorted().first()
+                allocatedEmulators[emulator] = ref
+                return emulator
+            } catch (e: RuntimeException) {
```

Comment:
Catch expected exception of `NoSuchElementException` only and rethrow as OverCapacityException.
If any unexpected exception happens, let's pass it through.

Reply:
ok